File: app/src/main/java/net/ddns/myapplication/SetAlarmActivity.java
package net.ddns.myapplication;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.DialogFragment;

import android.Manifest;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.MediaStore;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.webkit.MimeTypeMap;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.NumberPicker;
import android.widget.SeekBar;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.TimePicker;
import android.widget.Toast;
import android.widget.ToggleButton;

import net.ddns.myapplication.fragment.TimeRepeatDialogFragment;
import net.ddns.myapplication.table.NormalAlarm;
import net.ddns.myapplication.table.Song;
import net.ddns.myapplication.table.TimeRepeat;
import net.ddns.myapplication.table.Vibration;

import java.sql.Date;
import java.sql.Time;
import java.text.Format;
import java.text.SimpleDateFormat;

public class SetAlarmActivity extends AppCompatActivity implements TimeRepeatDialogFragment.OnCompleteListener {
    ImageButton backBtn;
    TextView txtvTime;
    TextView txtvWeek;
    TextView txtvAlarmTitle;
    TextView txtvAlarmSound;
    TextView txtvAlarmVibration;
    TextView txtvAlarmAgain;
    TimePicker tpSelectTime;
    ToggleButton[] tbtnWeek = new ToggleButton[7];
    LinearLayout llAlarmTitle;
    LinearLayout llAlarmSound;
    LinearLayout llAlarmVibration;
    LinearLayout llAlarmAgain;
    Switch switchAlarmTitle;
    Switch switchAlarmSound;
    Switch switchAlarmVibration;
    Switch switchAlarmAgain;
    Button btnAlarmDelete;
    Button btnAlarmSave;
    Button btnWeekday;
    Button btnWeekend;
    Button btnAllday;
    SeekBar seekBarSound;
    AlarmDatabase db;
    MediaPlayer mediaPlayer;
    Intent intent;

    Song selectedSong;
    Vibration selectedVibration;
    TimeRepeat selectedTimeRepeat;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_set_alarm);

        db = AlarmDatabase.getAppDatabase(this);

//        //UI 갱신 (라이브데이터 Observer 이용, 해당 디비값이 변화가생기면 실행됨)
//        db.normalAlarmDao().getAll().observe(this, new Observer<List<NormalAlarm>>() {
//            @Override
//            public void onChanged(List<NormalAlarm> normalAlarms) {
//                view.setText(normalAlarms.toString());
//            }
//        });

        findId();
        setDefaultTime();
        setListener();
    }

    private void findId(){
        backBtn = (ImageButton)findViewById(R.id.imgBtnBack);
        txtvTime = (TextView)findViewById(R.id.txtvTime);
        txtvWeek = (TextView)findViewById(R.id.txtvWeek);
        txtvAlarmTitle = (TextView)findViewById(R.id.txtvAlarmTitle);
        txtvAlarmSound = (TextView)findViewById(R.id.txtvAlarmSound);
        txtvAlarmVibration = (TextView)findViewById(R.id.txtvAlarmVibration);
        txtvAlarmAgain = (TextView)findViewById(R.id.txtvAlarmAgain);
        tpSelectTime = (TimePicker)findViewById(R.id.tpSelectTime);
        tbtnWeek[0] = (ToggleButton)findViewById(R.id.tbtnSun);
        tbtnWeek[1] = (ToggleButton)findViewById(R.id.tbtnMon);
        tbtnWeek[2] = (ToggleButton)findViewById(R.id.tbtnTue);
        tbtnWeek[3] = (ToggleButton)findViewById(R.id.tbtnWed);
        tbtnWeek[4] = (ToggleButton)findViewById(R.id.tbtnThu);
        tbtnWeek[5] = (ToggleButton)findViewById(R.id.tbtnFri);
        tbtnWeek[6] = (ToggleButton)findViewById(R.id.tbtnSat);
        llAlarmTitle = (LinearLayout)findViewById(R.id.llAlarmTitle);
        llAlarmSound = (LinearLayout)findViewById(R.id.llAlarmSound);
        llAlarmVibration = (LinearLayout)findViewById(R.id.llAlarmVibration);
        llAlarmAgain = (LinearLayout)findViewById(R.id.llAlarmAgain);
        switchAlarmTitle = (Switch)findViewById(R.id.switchAlarmTitle);
        switchAlarmSound = (Switch)findViewById(R.id.switchAlarmSound);
        switchAlarmVibration = (Switch)findViewById(R.id.switchAlarmVibration);
        switchAlarmAgain = (Switch)findViewById(R.id.switchAlarmAgain);
        btnAlarmDelete = (Button)findViewById(R.id.btnAlarmDelete);
        btnAlarmSave = (Button)findViewById(R.id.btnAlarmSave);
        btnWeekday = (Button)findViewById(R.id.btnWeekday);
        btnWeekend = (Button)findViewById(R.id.btnWeekend);
        btnAllday = (Button)findViewById(R.id.btnAllday);
        seekBarSound = (SeekBar)findViewById(R.id.seekBarSound);

        seekBarSound.setMax(15);
    }

    private void setDefaultTime(){
        long sNow = System.currentTimeMillis();
        Date dNow = new Date(sNow);
        Format formatter = new SimpleDateFormat("H:mm");
        String now = formatter.format(dNow);
        int h = Integer.parseInt(now.split(":")[0]);
        int m = Integer.parseInt(now.split(":")[1]);

        tpSelectTime.setCurrentHour(h);
        tpSelectTime.setCurrentMinute(m);
        txtvTime.setText(convTimeToString(tpSelectTime.getCurrentHour(), tpSelectTime.getCurrentMinute()));
        txtvWeek.setText(convWeekInfo());
    }

    private void setListener(){
        backBtn.setOnClickListener(clickListener);
        btnAlarmSave.setOnClickListener(clickListener);
        btnAlarmDelete.setOnClickListener(clickListener);
        btnWeekday.setOnClickListener(clickListener);
        btnWeekend.setOnClickListener(clickListener);
        btnAllday.setOnClickListener(clickListener);
        llAlarmTitle.setOnClickListener(clickListener);
        llAlarmSound.setOnClickListener(clickListener);
        llAlarmVibration.setOnClickListener(clickListener);
        llAlarmAgain.setOnClickListener(clickListener);
        tpSelectTime.setOnTimeChangedListener(timeChangedListener);
        seekBarSound.setOnSeekBarChangeListener(seekBarChangeListener);

        for(int i = 0; i < tbtnWeek.length; i++){
            tbtnWeek[i].setOnCheckedChangeListener(checkedChangeListener);
        }
    }

    private String convTimeToString(int hourOfDay, int minute){
        Time t = new Time(hourOfDay, minute, 0);
        Format formatter = new SimpleDateFormat("a h:mm");

        return formatter.format(t);
    }

    private String convWeekInfo(){
        String weekInfo = "";
        Boolean isAllDay = true;

        for(int i = 0; i < tbtnWeek.length; i++){
            if(tbtnWeek[i].isChecked()){
                weekInfo += tbtnWeek[i].getText();
            }else{
                isAllDay = false;
            }
        }

        weekInfo = isAllDay ? getRStr(R.string.all_day) : weekInfo;

        return weekInfo;
    }

    private void convWeekCheck(Boolean weekInfo[]){
        for(int i = 0; i < tbtnWeek.length; i++){
            tbtnWeek[i].setChecked(weekInfo[i]);
        }
    }

    private String getRStr(int id){
        return getResources().getString(id);
    }

    private String[] getAllMp3Path(){
        String[] resultPath = null;
        // 외장 메모리 접근 권한을 가지고 있는지 확인. ( Marshmallow 이상 )  // mAcitivity == Main Activity
        if(ContextCompat.checkSelfPermission(getApplicationContext(), Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
            String selectionMimeType = MediaStore.Files.FileColumns.MIME_TYPE + "=?";
            // 찾고자하는 파일 확장자명.
            String mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension("mp3");

            String[] selectionArgsMp3 = new String[]{mimeType};

            Cursor c = getContentResolver().query(
                    MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
                    new String[]{MediaStore.Audio.Media.DATA}, selectionMimeType, selectionArgsMp3, null);

            if (c.getCount() == 0)
                return null;

            resultPath = new String[c.getCount()];
            while (c.moveToNext()) {
                // 경로 데이터 셋팅.
                resultPath[c.getPosition()] = c.getString(c.getColumnIndexOrThrow(MediaStore.Audio.Media.DATA));
                Log.d("===TAG", resultPath[c.getPosition()]);
            }
        }
        return resultPath;
    }

    @Override
    public void onInputedData(String minute, String count) {
        txtvAlarmAgain.setText(
                minute+"분, "+count + (count.equals(getRStr(R.string.repeat_infinite))? "":"회")
        );

        selectedTimeRepeat = new TimeRepeat(
                Integer.parseInt(minute),
                count.equals(getRStr(R.string.repeat_infinite))?0:Integer.parseInt(count)
        );
        switchAlarmAgain.setChecked(true);
    }

    //    ===Listener===
    View.OnClickListener clickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            switch(v.getId()){
                case R.id.imgBtnBack:
                    finish();
                    break;
                case R.id.btnAlarmDelete:
                    finish();
                    break;
                case R.id.btnAlarmSave:
//                    //DB에 데이터 INSERT
//                    if(txtvAlarmTitle.getText().toString().trim().length() <= 0) {
//                        Toast.makeText(getApplicationContext(), "no title", Toast.LENGTH_SHORT).show();
//                    }else{
//                        new InsertAsyncTask (db.normalAlarmDao()).execute(new NormalAlarm(txtvAlarmTitle.getText().toString().trim(), "time", 1));
//                    }
                    finish();
                    break;
                case R.id.btnWeekday:
                    convWeekCheck( new Boolean[] {false, true,true,true,true,true, false});
                    break;
                case R.id.btnWeekend:
                    convWeekCheck( new Boolean[] {true, false,false,false,false,false, true});
                    break;
                case R.id.btnAllday:
                    convWeekCheck( new Boolean[] {true, true,true,true,true,true, true});
                    break;
                case R.id.llAlarmTitle:
                    final EditText editText = new EditText(getApplicationContext());
                    editText.setText(txtvAlarmTitle.getText().toString().equals(getRStr(R.string.default_sel)) ? null : txtvAlarmTitle.getText().toString());
                    AlertDialog.Builder setTitleDialog = new AlertDialog.Builder(SetAlarmActivity.this);
                    setTitleDialog.setTitle(getRStr(R.string.alarm_title));
                    setTitleDialog.setView(editText);

                    setTitleDialog.setPositiveButton(getRStr(R.string.ok), new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            if(editText.getText().toString().equals("")){
                                txtvAlarmTitle.setText(getRStr(R.string.default_sel));
                                switchAlarmTitle.setChecked(false);
                                Toast.makeText(getApplicationContext(), "알람 제목을 설정하기 않았습니다.", Toast.LENGTH_SHORT).show();
                            }else{
                                new InsertAsyncTask(db.normalAlarmDao()).execute(
                                        new NormalAlarm(
                                                editText.getText().toString(),
                                                convTimeToString(tpSelectTime.getCurrentHour(), tpSelectTime.getCurrentMinute()),
                                                11
                                        )
                                );

                                txtvAlarmTitle.setText(editText.getText().toString());
                                switchAlarmTitle.setChecked(true);
                            }
                        }
                    });
                    setTitleDialog.show();

                    break;
                case R.id.llAlarmSound:
                    intent = new Intent(getApplicationContext(), SoundSelectActivity.class);
                    startActivityForResult(intent, 3000);
                    break;
                case R.id.llAlarmVibration:
                    intent = new Intent(getApplicationContext(), VibrationSelectActivity.class);
                    startActivityForResult(intent, 3001);
                    break;
                case R.id.llAlarmAgain:
                    DialogFragment timeRepeatDialogFragment = new TimeRepeatDialogFragment();
                    timeRepeatDialogFragment.show(getSupportFragmentManager(),"dialog");

//                    intent = new Intent(getApplicationContext(), TimeRepeatActivity.class);
//                    startActivityForResult(intent, 3002);
                    break;
            }
        }
    };

    TimePicker.OnTimeChangedListener timeChangedListener = new TimePicker.OnTimeChangedListener(){
        @Override
        public void onTimeChanged(TimePicker view, int hourOfDay, int minute) {
            txtvTime.setText( convTimeToString(hourOfDay, minute) );
        }
    };

    CompoundButton.OnCheckedChangeListener checkedChangeListener = new CompoundButton.OnCheckedChangeListener() {
        @Override
        public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
            if(buttonView.getText().equals(getRStr(R.string.sun)) && !isChecked){
                buttonView.setTextColor(getResources().getColor(R.color.red));
            }else if(buttonView.getText().equals(getRStr(R.string.sat)) && !isChecked){
                buttonView.setTextColor(getResources().getColor(R.color.blue));
            }else if(isChecked){
                buttonView.setTextColor(getResources().getColor(R.color.main));
            }else{
                buttonView.setTextColor(getResources().getColor(R.color.black));
            }

            txtvWeek.setText(convWeekInfo());
        }
    };

    SeekBar.OnSeekBarChangeListener seekBarChangeListener = new SeekBar.OnSeekBarChangeListener() {
        @Override
        public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            switch(seekBar.getId()){
                case R.id.seekBarSound:
                    AudioManager audioManager = (AudioManager)getSystemService(AUDIO_SERVICE);
                    audioManager.setStreamVolume(AudioManager.STREAM_RING, progress, 0);
                    break;
            }
        }

        @RequiresApi(api = Build.VERSION_CODES.O)
        @Override
        public void onStartTrackingTouch(SeekBar seekBar) {
            switch(seekBar.getId()){
                case R.id.seekBarSound:
                    mediaPlayer = new MediaPlayer();

                    if(selectedSong == null){
                        Toast.makeText(getApplicationContext(), getRStr(R.string.sound_not_set_message), Toast.LENGTH_SHORT).show();
                        break;
                    }
                    try{
                        mediaPlayer.setDataSource(getApplicationContext(), Uri.parse(selectedSong.getUri()));
                        mediaPlayer.setAudioStreamType(AudioManager.STREAM_RING);
                        mediaPlayer.setLooping(true);
                        mediaPlayer.prepare();
                        mediaPlayer.start();
                    }catch(Exception e){
                        mediaPlayer.pause();
                        mediaPlayer.release();
                        mediaPlayer = new MediaPlayer();
                        Log.e("seekbar error", e.toString());
                    }
                    break;
            }
        }

        @Override
        public void onStopTrackingTouch(SeekBar seekBar) {
            switch(seekBar.getId()){
                case R.id.seekBarSound:
                    if(mediaPlayer != null && mediaPlayer.isPlaying()){
                        Log.d("ring play!!!", "");
                        mediaPlayer.stop();
                        mediaPlayer.release();
                    }
                    mediaPlayer.release();
                    break;
            }
        }
    };


//    ===ActivityResult===
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        switch(resultCode){
            case RESULT_OK:
                switch(requestCode){
                    case 3000:
                        setSoundSelect((Song)data.getSerializableExtra("song"));
                        break;
                    case 3001:
                        setVibrationSelect((Vibration)data.getSerializableExtra("vibration"));
                        break;
                    default:

                        break;
                }
                break;
            case RESULT_CANCELED:
                switch (requestCode){
                    case 3000:
                        setSoundSelect(selectedSong);
                        break;
                    case 3001:
                        setVibrationSelect(selectedVibration);
                        break;
                    default:
                        break;
                }
                break;
            default:
                switch(requestCode){
                    case 3000:
                        Log.e("requesCode:3000", "sound select error");
                        break;
                    case 3001:
                        Log.e("requesCode:3001", "vibration select error");
                        break;
                    default:

                        break;
                }
                break;
        }
    }

    private void setSoundSelect(Song s){
        selectedSong = s;
        txtvAlarmSound.setText(s != null? s.getTitle():getRStr(R.string.default_sel));
        switchAlarmSound.setChecked(s != null);
    }

    private void setVibrationSelect(Vibration vib){
        selectedVibration = vib;
        txtvAlarmVibration.setText(vib != null ? vib.getName():getRStr(R.string.default_sel));
        switchAlarmVibration.setChecked(vib != null);
    }
}

File: app/src/main/java/net/ddns/yline/alarm/SetAlarmActivity.kt
package net.ddns.yline.alarm

import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.AudioManager.STREAM_RING
import android.media.MediaPlayer
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.text.Editable
import android.util.Log
import android.view.View
import android.widget.*
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.ContextCompat
import io.reactivex.rxjava3.disposables.CompositeDisposable
import net.ddns.yline.alarm.databinding.ActivitySetAlarmBinding
import net.ddns.yline.alarm.fragment.RepeatDialogFragment
import net.ddns.yline.alarm.table.Song
import net.ddns.yline.alarm.table.Vibration
import java.lang.Exception
import java.sql.Time
import java.text.SimpleDateFormat
import java.util.*

class SetAlarmActivity : AppCompatActivity() {
    private val binding by lazy { ActivitySetAlarmBinding.inflate(layoutInflater) }
    private var selectedSound: Song? = null
    private var selectedVibration: Vibration? = null
    private val myCompositeDisposable = CompositeDisposable()
    private lateinit var mediaPlayer: MediaPlayer

    // ===== override function =====
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        setDefaultTime()
        setListener()
    }

    override fun onDestroy() {
        this.myCompositeDisposable.clear()
        super.onDestroy()
    }

    // ===== function =====
    private fun setDefaultTime(){
        val now:String = SimpleDateFormat("H:mm").format(Date(System.currentTimeMillis()))
        val h:Int = now.substring(0, now.indexOf(":")).toInt()
        val m:Int = now.substring(now.indexOf(":") + 1).toInt()

        binding.timepickerTimeSet.apply {
            hour = h
            minute = m
        }
        binding.textviewTopTime.text = convTimeToString(h, m)
        binding.textviewTopWeek.text = convWeekInfo()
    }

    private fun setListener(){
        with(binding) {
            ClickListener().also {
                imageButtonBack.setOnClickListener(it)
                buttonAlarmSave.setOnClickListener(it)
                buttonAlarmDelete.setOnClickListener(it)
                buttonTimeAll.setOnClickListener(it)
                buttonTimeWeekday.setOnClickListener(it)
                buttonTimeWeekend.setOnClickListener(it)
                constraintNameSet.setOnClickListener(it)
                constraintSoundSet.setOnClickListener(it)
                constraintVibrationSet.setOnClickListener(it)
                constraintRepeatSet.setOnClickListener(it)
                switchVibration.setOnClickListener(it)
            }
            CheckedChangeListener().also {
                togglebuttonSelSun.setOnCheckedChangeListener(it)
                togglebuttonSelMon.setOnCheckedChangeListener(it)
                togglebuttonSelTue.setOnCheckedChangeListener(it)
                togglebuttonSelWed.setOnCheckedChangeListener(it)
                togglebuttonSelThu.setOnCheckedChangeListener(it)
                togglebuttonSelFri.setOnCheckedChangeListener(it)
                togglebuttonSelSat.setOnCheckedChangeListener(it)
            }

            timepickerTimeSet.setOnTimeChangedListener(TimeChangedListener())
            seekbarSound.setOnSeekBarChangeListener(SeekBarChangeListener())
        }
    }

    private fun convTimeToString(hourOfDay:Int, minute:Int):String{
        return SimpleDateFormat("a h:mm").format(Time(hourOfDay, minute, 0))
    }

    private fun convWeekInfo():String {
        return arrayOf(
            binding.togglebuttonSelSun, binding.togglebuttonSelMon, binding.togglebuttonSelTue,
            binding.togglebuttonSelWed, binding.togglebuttonSelThu, binding.togglebuttonSelFri,
            binding.togglebuttonSelSat).fold(""){ acc, item ->
            if(item.isChecked) acc + item.text
            else acc + ""
        }
    }

    private fun convWeekCheck(weekInfo:BooleanArray){
        arrayOf(
            binding.togglebuttonSelSun, binding.togglebuttonSelMon, binding.togglebuttonSelTue,
            binding.togglebuttonSelWed, binding.togglebuttonSelThu, binding.togglebuttonSelFri,
            binding.togglebuttonSelSat).forEachIndexed { idx, toggleButton -> toggleButton.isChecked = weekInfo[idx] }
    }

    private fun setSoundSelect(s:Song){
        selectedSound = s
        binding.textviewSoundTitle.text = s.title
        binding.switchSound.isChecked = s != null
    }

    private fun setVibrationSelect(vib: Vibration){
        selectedVibration = vib
        binding.textviewVibrationTitle.text = vib.name
        binding.switchVibration.isChecked = vib != null
    }

    private fun pretreatmentAlarmTitle(title:String):String{
        return title.replace("\n", "").run {
            if(this.length > 20) this.substring(0, 20) else this
        }
    }

    private fun vibrationSwitchAction(){
//        ===test===
//        val switchChangeObservable = binding.switchVibration.checkedChanges()
//        val switchChangeSubscription:Disposable =
//            switchChangeObservable.debounce( 800, TimeUnit.MICROSECONDS )
//                .subscribeOn(Schedulers.io())
//                .subscribeBy {
//
//                }
//        myCompositeDisposable.add(switchChangeSubscription)

        binding.run {
            if(textviewVibrationTitle.text == resources.getString(R.string.textview_default_select) && switchVibration.isChecked){
                switchVibration.isChecked = false
                toast("선택된 진동이 없습니다.")
                Log.d("vibration selection null", "SetAlarmActivity/vibrationSwitchAction")
            }
        }
    }

    private fun soundSetAction(){
        startForSoundResult.launch(
            Intent(applicationContext, SoundSelectActivity::class.java).apply {
                putExtra("selectedSound", if(selectedSound != null) selectedSound else null)
            }
        )
    }

    private fun vibrationSetAction(){
        startForVibrationResult.launch(
            Intent(applicationContext, VibrationSelectActivity::class.java).apply {
                putExtra("selectedVibration", if(selectedVibration != null) selectedVibration else null)
            }
        )
    }

    private fun repeatSetAction(){
        RepeatDialogFragment().apply {
            setButtonClickListener(object:RepeatDialogFragment.OnCompleteListener{
                override fun onButtonRepeatSetClicked(minute: String?, count: String?) {
                    binding.apply {
                        textviewRepeatTitle.text = String.format(getString(R.string.textview_repeat_text), minute, count)
                        switchRepeat.isChecked = true
                    }
                }
            })
        }.show(supportFragmentManager, "SetRepeatDialog")
    }

    private fun alarmSaveAction(){
        toast("save!!!")
        finish()
    }

    private fun clickDefaultAction(v:View?){
        Log.e("|||null listener error", v.toString())
        toast("error")
    }

    private fun repeatSwitchAction(){

    }

    private fun setTitleAction(){
        val editText:EditText = EditText(applicationContext).apply {
            text = binding.textviewNameTitle.text.let {
                if(it == resources.getString(R.string.textview_default_select)) null
                else it.toString().toEditable()
            }
        }
        AlertDialog.Builder(this@SetAlarmActivity).apply {
            setTitle(resources.getString(R.string.alarm_name))
            setMessage("20자 이내로 알람 이름을 설정해주세요.")
            setView(editText)
            setPositiveButton(resources.getString(R.string.ok)) { _, _ ->
                if (editText.text.toString() == "") {
                    binding.textviewNameTitle.text = resources.getString(R.string.textview_default_select)
                    binding.switchName.isChecked = false
                    toast("알람 제목을 설정하기 않았습니다.")
                } else {
                    binding.textviewNameTitle.text = pretreatmentAlarmTitle(editText.text.toString())
                    binding.switchName.isChecked = true
                    Log.d("|||text len", binding.textviewNameTitle.text.length.toString())
                }
            }
        }.show()
    }


    // ===== listener =====
    inner class ClickListener:View.OnClickListener{
        override fun onClick(v: View?) {
            binding.run {
                when(v?.id){
                    imageButtonBack.id -> finish()
                    buttonAlarmDelete.id -> finish()
                    buttonAlarmSave.id -> alarmSaveAction()
                    buttonTimeAll.id -> convWeekCheck(booleanArrayOf(true, true, true, true, true, true, true))
                    buttonTimeWeekday.id -> convWeekCheck(booleanArrayOf(false, true, true, true, true, true, false))
                    buttonTimeWeekend.id -> convWeekCheck(booleanArrayOf(true, false, false, false, false, false, true))
                    constraintNameSet.id -> setTitleAction()
                    constraintSoundSet.id -> soundSetAction()
                    constraintVibrationSet.id -> vibrationSetAction()
                    constraintRepeatSet.id -> repeatSetAction()
                    switchVibration.id -> vibrationSwitchAction()
                    switchRepeat.id -> repeatSwitchAction()
                    else -> clickDefaultAction(v)
                }
            }
        }
    }

    inner class CheckedChangeListener:CompoundButton.OnCheckedChangeListener{
        override fun onCheckedChanged(buttonView: CompoundButton?, isChecked: Boolean) {
            when{
                buttonView == null -> toast("error")
                buttonView.text == "일" && !isChecked -> buttonView.setTextColor(ContextCompat.getColor(applicationContext, R.color.red))
                buttonView.text == "토" && !isChecked -> buttonView.setTextColor(ContextCompat.getColor(applicationContext, R.color.blue))
                isChecked -> buttonView.setTextColor(ContextCompat.getColor(applicationContext, R.color.main))
                else -> buttonView.setTextColor(ContextCompat.getColor(applicationContext, R.color.black))
            }
            binding.textviewTopWeek.text = convWeekInfo()
        }
    }

    inner class SeekBarChangeListener:SeekBar.OnSeekBarChangeListener{
        override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
            (getSystemService(AUDIO_SERVICE) as AudioManager).apply {
                setStreamVolume(STREAM_RING, progress, 0)
            }
        }

        override fun onStartTrackingTouch(seekBar: SeekBar) {
            try {
                if(selectedSound == null){
                    toast("설정된 소리가 없습니다.")
                    return
                }
                mediaPlayer = MediaPlayer()
                mediaPlayer.apply {
                    setDataSource(applicationContext, Uri.parse(selectedSound!!.uri))
                    setAudioAttributes(AudioAttributes.Builder().setLegacyStreamType(STREAM_RING).build())
                    isLooping = true
                    prepare()
                    start()
                }
            }catch (e:Exception){
                mediaPlayer.pause()
                mediaPlayer.release()
                mediaPlayer = MediaPlayer()
                Log.e("media seekbar error", e.stackTraceToString())
            }
        }

        override fun onStopTrackingTouch(seekBar: SeekBar) {
            mediaPlayer.run {
                if(this != null && isPlaying){
                    stop()
                    release()
                }
                release()
            }
        }
    }

    inner class TimeChangedListener:TimePicker.OnTimeChangedListener{
        override fun onTimeChanged(v: TimePicker?, hourOfDay: Int, minute: Int) {
            binding.textviewTopTime.text = convTimeToString(hourOfDay, minute)
        }
    }


    // ===== activity result =====
    private val startForSoundResult = registerForActivityResult(ActivityResultContracts.StartActivityForResult()){ result: ActivityResult ->
        when(result.resultCode){
            RESULT_OK -> {
                selectedSound = result.data?.getSerializableExtra("selectedSound") as Song
                if(selectedSound == null){
                    binding.switchSound.isChecked = false
                    Log.e("sound is null", "SetAlarmActivity/startForSoundResult")

                    toast("error")
                }else{
                    binding.apply {
                        textviewSoundTitle.text = selectedSound!!.title
                        switchSound.isChecked = true
                    }
                    toast("success")
                }
            }
            RESULT_CANCELED -> {
                Log.d("sound set canceled", "SetAlarmActivity/startForSoundResult")
            }
        }
    }

    private val startForVibrationResult = registerForActivityResult(ActivityResultContracts.StartActivityForResult()){ result: ActivityResult ->
        when(result.resultCode){
            RESULT_OK -> {
                selectedVibration = result.data?.getSerializableExtra("selectedVibration") as Vibration
                if(selectedVibration == null){
                    binding.switchVibration.isChecked = false
                    Log.e("vibration is null", "SetAlarmActivity/startForVibrationResult")
                    toast("error")
                }else{
                    binding.apply {
                        textviewVibrationTitle.text = selectedVibration!!.name
                        switchVibration.isChecked = true
                    }
                    toast("success")
                }
            }
            RESULT_CANCELED -> {
                Log.d("vibration set canceled", "SetAlarmActivity/startForVibrationResult")
            }
        }
    }

    // ===== extension function =====
    private fun String.toEditable() = Editable.Factory.getInstance().newEditable(this)

    // ===== test function =====
    private fun toast(s:String){
        val str = when(s){
            "success" -> getString(R.string.setting_complete)
            "error" -> getString(R.string.setting_error)
            else -> s
        }
        Toast.makeText(applicationContext, str, Toast.LENGTH_SHORT).show()
    }

    // ===== getting ready =====
//    private fun setCalendarDialog(){
//        var listener = DatePickerDialog.OnDateSetListener { datePicker, i, i2, i3 ->
//            val test = "${i}년 ${i2 + 1}월 ${i3}일 ${SimpleDateFormat("EEEE").format(Date(i, i2, i3-1))}"
//            toast(test)
//        }
//
//        Calendar.getInstance().apply {
//            DatePickerDialog(this@SetAlarmActivity, listener, get(Calendar.YEAR), get(Calendar.MONTH), get(Calendar.DAY_OF_MONTH)).show()
//
//        }
//    }
}